waypointy/oc/xml2sql.php

```diff
@@ -1,166 +1,345 @@
-$BAZY_OC['us']['szukaj'] = 'http://www.opencaching.us/searchplugin.php?userinput=';
+    foreach ($changes as $change) {
 
-// $BAZY_OC['jp']['prefix'] = 'OJ';
-// $BAZY_OC['jp']['url'] = "http://www.opencaching.jp/xml/ocxml11.php?cache=1&session=0&charset=utf-8&cdata=1&xmldecl=1&ocxmltag=1&doctype=0&zip=0&modifiedsince=";
-// $BAZY_OC['jp']['szukaj'] = 'http://www.opencaching.jp/searchplugin.php?userinput=';
+        if ($change->object_type != 'geocache') {
+            continue;
+        }
 
+        $id = $change->object_key->code;
 
-// $BAZY_OC['no']['prefix'] = 'OS';
-// $BAZY_OC['no']['url'] = "http://www.opencaching.no/xml/ocxml11.php?cache=1&session=0&charset=utf-8&cdata=1&xmldecl=1&ocxmltag=1&doctype=0&zip=0&modifiedsince=";
-// $BAZY_OC['no']['szukaj'] = 'http://www.opencaching.no/searchplugin.php?userinput=';
+        if ($change->change_type == 'delete') {
+            //delete from DB
 
+            $sql = 'DELETE FROM `gk-waypointy` WHERE waypoint="' . mysqli_real_escape_string($link, $id) . '"';
 
-$BAZY_OC['nl']['prefix'] = 'OB';
-$BAZY_OC['nl']['url'] = "http://www.opencaching.nl/xml/ocxml11.php?cache=1&session=0&charset=utf-8&cdata=1&xmldecl=1&ocxmltag=1&doctype=0&zip=0&modifiedsince=";
-$BAZY_OC['nl']['szukaj'] = 'http://www.opencaching.nl/searchplugin.php?userinput=';
+            $result = mysqli_query($link, $sql);
 
+            if ($result === false) { // ooooPs we got an import error !
+                print("error sql : id:$id - query:$sql - mysqli_error:" . mysqli_error($link) . "\n");
+            }
+            $nDeleted++;
+            continue;
+        }
 
-/*
-// tylko odkomentować //
-$BAZY_OC['cz']['prefix'] = 'OZ';
-$BAZY_OC['cz']['url'] = "/home/geokrety/public_html/tools/oc-cz.xml";
-$BAZY_OC['cz']['szukaj'] = 'http://www.opencaching.cz/searchplugin.php?userinput=';
-*/
+        // Check for needed fields and make an update
+        $sqlInsert = array();
+        $sqlValues = array();
+        $sqlUpdate = array();
 
+        if (isset($change->data->names)) {
+            $name = mysqli_real_escape_string($link, implode(' | ', (array)$change->data->names));
+            $sqlInsert [] = 'name';
+            $sqlValues [] = "'$name'";
+            $sqlUpdate [] = "name='$name'";
+        }
+        if (isset($change->data->owner->username)) {
+            $owner = mysqli_real_escape_string($link, (string)$change->data->owner->username);
+            $sqlInsert [] = 'owner';
+            $sqlValues [] = "'$owner'";
+            $sqlUpdate [] = "owner='$owner'";
+        }
+        if (isset($change->data->location)) {
+            $location = explode('|', $change->data->location);
+            $lon = mysqli_real_escape_string($link, (string)$location[0]);
+            $lat = mysqli_real_escape_string($link, (string)$location[1]);
+
+            $sqlInsert [] = 'lon';
+            $sqlValues [] = "'$lon'";
```

Comment:
Probably not, will fix.

Though I'm not really sure how to use prepared statement for this type of logic when I compose the whole query on the fly, when amount of fields depends on the input

Reply:
done. However for especially this query I don't really like idea of using anonymous prepared statement (`?`) because it will make the code much less readable IMO. SO left the query as is

Reply:
Security over readability please 🤲. We should not trust content even if it come from "partners".

Reply:
https://stackoverflow.com/a/15786451/944936

Reply:
There is de-facto no difference between using prepared statement and just proper escaping preventing SQL injections here. Security is completely the same. But whatever, changed.